packages/core/test/new-api/readEventArgument.ts

```diff
@@ -0,0 +1,211 @@
+import { assert } from "chai";
+
+import { FutureType, ReadEventArgumentFuture } from "../../src";
+import { defineModule } from "../../src/new-api/define-module";
+import { ModuleConstructor } from "../../src/new-api/internal/module-builder";
+
+describe("Read event argument", () => {
+  describe("creating modules with it", () => {
+    it("Should support reading arguments from all the futures that can emit them", () => {
+      const fakeArtifact = {} as any;
+
+      const defintion = defineModule("Module1", (m) => {
+        const contract = m.contract("Contract");
+        const contractFromArtifact = m.contractFromArtifact(
+          "ContractFromArtifact",
+          fakeArtifact
+        );
+        const call = m.call(contract, "fuc");
+
+        m.readEventArgument(contract, "EventName1", "arg1");
+        m.readEventArgument(contractFromArtifact, "EventName2", "arg2");
+        m.readEventArgument(call, "EventName3", "arg3");
+
+        return { contract, contractFromArtifact };
+      });
+
+      const constructor = new ModuleConstructor(0, []);
+      const mod = constructor.construct(defintion);
+
+      const callFuture = Array.from(mod.futures).find(
+        (f) => f.type === FutureType.NAMED_CONTRACT_CALL
+      );
+
+      const [read1, read2, read3] = Array.from(mod.futures).filter(
+        (f) => f.type === FutureType.READ_EVENT_ARGUMENT
+      ) as ReadEventArgumentFuture[];
+
+      assert.equal(read1.futureToReadFrom, mod.results.contract);
+      assert.equal(read2.futureToReadFrom, mod.results.contractFromArtifact);
+      assert.equal(read3.futureToReadFrom, callFuture);
+    });
+
+    it("Should infer the emitter from the future correctly", () => {
+      const defintion = defineModule("Module1", (m) => {
+        const contract = m.contract("Contract");
+        const call = m.call(contract, "fuc");
+
+        m.readEventArgument(contract, "EventName1", "arg1");
+        m.readEventArgument(call, "EventName2", "arg2");
+
+        return { contract };
+      });
+
+      const constructor = new ModuleConstructor(0, []);
+      const mod = constructor.construct(defintion);
+
+      const [read1, read2] = Array.from(mod.futures).filter(
+        (f) => f.type === FutureType.READ_EVENT_ARGUMENT
+      ) as ReadEventArgumentFuture[];
+
+      assert.equal(read1.emitter, mod.results.contract);
+      assert.equal(read2.emitter, mod.results.contract);
+    });
+
+    it("Should accept an explicit emitter", () => {
+      const defintion = defineModule("Module1", (m) => {
+        const contract = m.contract("Contract");
+        const call = m.call(contract, "fuc");
+
+        const contract2 = m.contract("Contract2");
+
+        m.readEventArgument(contract, "EventName1", "arg1");
+        m.readEventArgument(call, "EventName2", "arg2", { emitter: contract2 });
```

Comment:
When do I want to change the contract, can you walk me through the use case?

Reply:
The answer to this is that a call to a contract can delegate on to a whole chain of further contract function calls. Hence we need a way of saying which contract in the call chain is the emitter, not necessarily the top level contract that was called.